File: ts/project/.sdk/tm/c/utility/prepare_params.c
// prepare_params utility (mirrors utility/prepare_params.rs).

#include "sdk.h"

voxgig_value* prepare_params_util(Context* ctx) {
  voxgig_value* point = ctx->point;

  voxgig_value* args = to_map(getp(point, "args"));
  voxgig_value* params = voxgig_new_list();
  if (voxgig_is_map(args)) {
    voxgig_value* p = getp(args, "params");
    if (voxgig_is_list(p)) params = p;
  }

  voxgig_value* out = voxgig_new_map();
  if (voxgig_is_list(params)) {
    voxgig_list* pl = voxgig_as_list(params);
    for (size_t i = 0; i < pl->len; i++) {
      voxgig_value* pd = pl->items[i];
      voxgig_value* val = param_util(ctx, pd);
      if (!v_is_noval(val) && !v_is_null(val) && voxgig_is_map(pd)) {
        const char* name = get_str(pd, "name");
        if (name && name[0] != '\0') {
          setp(out, name, v_share(val));
        }
      }
    }
  }
  return out;
}

File: ts/project/.sdk/tm/rust/utility/prepare_params.rs
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::helpers::{get_str, getp, setp, to_map};
use crate::utility::voxgigstruct::Value;

pub fn prepare_params_util(ctx: &Rc<Context>) -> Value {
    let point = ctx.point.borrow().clone();

    let params = match to_map(&getp(&point, "args")) {
        Value::Map(_) => match getp(&to_map(&getp(&point, "args")), "params") {
            Value::List(l) => Value::List(l),
            _ => Value::empty_list(),
        },
        _ => Value::empty_list(),
    };

    let out = Value::empty_map();
    if let Value::List(pl) = &params {
        for pd in pl.borrow().iter() {
            let val = crate::utility::param::param_util(ctx, pd);
            if !val.is_noval() && !val.is_null() {
                if let Value::Map(_) = pd {
                    if let Some(name) = get_str(pd, "name").filter(|n| !n.is_empty()) {
                        setp(&out, &name, val);
                    }
                }
            }
        }
    }

    out
}
